sonarts-core/src/utils/sonar-analysis.ts

```diff
@@ -0,0 +1,132 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import * as ts from "typescript";
+import { lineAndCharacter } from "./navigation";
+import { toSonarLine } from "../runner/sonar-utils";
+import { TreeVisitor } from "./visitor";
+import * as tslint from "tslint";
+
+export class SonarRuleVisitor extends TreeVisitor {
+  private issues: SonarIssue[] = [];
+
+  public constructor(private ruleName: string) {
+    super();
+  }
+
+  public getIssues(): tslint.RuleFailure[] {
+    return this.issues;
+  }
+
+  public addIssue(node: ts.Node, message: string): SonarIssue {
+    const issue = new SonarIssue(new IssueLocation(node, message), this.ruleName);
+    this.issues.push(issue);
+    return issue;
+  }
+
+  public addIssueAtLocation(primaryLocation: IssueLocation): SonarIssue {
+    const issue = new SonarIssue(primaryLocation, this.ruleName);
+    this.issues.push(issue);
+    return issue;
+  }
+}
+
+export class TypedSonarRuleVisitor extends SonarRuleVisitor {
+  public constructor(ruleName: string, protected program: ts.Program) {
+    super(ruleName);
+  }
+}
+export class IssueLocation {
+  public readonly startLine: number;
+  public readonly startColumn: number;
+  public readonly endLine: number;
+  public readonly endColumn: number;
+
+  public constructor(public readonly node: ts.Node, public message?: string, public readonly lastNode: ts.Node = node) {
+    this.node = node;
+    this.message = message;
+
+    const startPosition = lineAndCharacter(node.getStart(), node.getSourceFile());
+    const endPosition = lineAndCharacter(lastNode.getEnd(), node.getSourceFile());
+
+    this.startLine = toSonarLine(startPosition.line);
+    this.startColumn = startPosition.character;
+    this.endLine = toSonarLine(endPosition.line);
+    this.endColumn = endPosition.character;
+  }
+
+  public toJson() {
+    return {
+      startLine: this.startLine,
+      startColumn: this.startColumn,
+      endLine: this.endLine,
+      endColumn: this.endColumn,
+    };
+  }
+}
+
+export class SonarIssue extends tslint.RuleFailure {
+  private cost?: number;
+  public readonly primaryLocation: IssueLocation;
+  private secondaryLocations: IssueLocation[] = [];
+
+  public constructor(primaryLocation: IssueLocation, ruleName: string) {
+    super(
+      primaryLocation.node.getSourceFile(),
+      primaryLocation.node.getStart(),
+      primaryLocation.lastNode.getEnd(),
+      primaryLocation.message!,
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonar-issue&issues=AWCdkH8rqDrCo5ysf9j0&open=AWCdkH8rqDrCo5ysf9j0)

[ISSUE_KEY]: AWCdkH8rqDrCo5ysf9j0

Reply:
This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED